qkeras/quantizers.py

```diff
@@ -439,9 +440,537 @@ def trainable_variables(self):
+        raise TypeError(
+            f"alpha must be, a string, an array, or None, not {type(alpha)}")
+
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def scale(self):
+    return self.quantization_scale / self.data_type_scale
+  
+  @property
+  def quantization_scale(self):
+    return self._quantization_scale
+  
+  @quantization_scale.setter
+  def quantization_scale(self, quantization_scale):
+    self._quantization_scale = quantization_scale
+    
+  @property
+  def data_type_scale(self):
+    """Quantization scale for the data type"""
+    # integer is sometimes cast as int32, so cast to float32 to avoid errors
+    integer = tf.cast(self.integer, tf.float32)
+    return K.pow(2.0, integer - self.bits + self.keep_negative)
+
+  @property
+  def auto_alpha(self):
+    """Returns true if using a data-dependent alpha"""
+
+    return isinstance(self.alpha, six.string_types)
+
+  @property
+  def use_sign_function(self):
+    """Return true if using sign function for quantization"""
+
+    return (self.bits == 1.0) and self.keep_negative
+
+  @property
+  def clip_bounds(self):
+    """Get bounds of clip range"""
+
+    if self.use_sign_function:
+      clip_min = K.cast_to_floatx(-0.5)
+      clip_max = K.cast_to_floatx(0.5)
+    else:
+      unsigned_bits_po2 = K.pow(2.0, self.bits - self.keep_negative)
+      # if symmetric, clip_min is negative of clip_max. Otherwise clip_min is
+      # lowered by 1, giving us one more representable number
+      clip_min = self.keep_negative * (-unsigned_bits_po2 + self.symmetric)
+      clip_max = unsigned_bits_po2 - K.cast_to_floatx(1.0)
+
+    return clip_min, clip_max
+  
+  def _set_default_quantization_scale(self):
+    """Calculate and set quantization_scale default"""
+    assert (
+        self._initialized
+    ), "Must initialize before calling _set_default_quantization_scale"
+
+
+    err_msg = (f"Bit count {self.bits} must exceed "
+              f" {self.integer + self.keep_negative}")
+    if self.bits < self.integer + self.keep_negative:
+      raise ValueError(err_msg)
+
+    # Set default quantization scale
+    self.quantization_scale = self.data_type_scale
+
+    # Set scales for tensor alpha
+    if self.alpha is not None and not self.auto_alpha:
+        self.quantization_scale = self.alpha * self.data_type_scale
+
+  def __call__(self, x):
+    """Core quantization function"""
+
+    # Build if not already built
+    self._build()
+
+    # Data type conversion
+    x = K.cast_to_floatx(x)
+    
+    if self.auto_alpha:
+      # get data-dependent quantization scale
+      quantization_scale = self._get_quantization_scale(x)
+    else:
+      # quantization scale determined by quantizer params, not data
+      # see _set_default_quantization_scale for more info
+      quantization_scale = self.quantization_scale
+
+    scaled_xq = self._scale_clip_and_round(x, quantization_scale)
+    xq = scaled_xq * quantization_scale
```

Comment:
This is very clear and easy to follow. Nice!

Reply:
Thanks!